doc/DESIGN.md

```diff
@@ -0,0 +1,1625 @@
+
+The CFG must have *no critical edges*. A critical edge is an edge from
+block A to block B such that A has more than one successor *and* B has
+more than one predecessor. For this definition, the entry block has an
+implicit predecessor, and any block that ends in a return has an
+implicit successor.
+
+Note that there are *no* requirements related to the ordering of
+blocks, and there is no requirement that the control flow be
+reducible. Some *heuristics* used by the allocator will perform better
+if the code is reducible and ordered in reverse postorder (RPO),
+however: in particular, (1) this interacts better with the
+contiguous-range-of-instruction-indices live range representation that
+we use, and (2) the "approximate loop depth" metric will actually be
+exact if both these conditions are met.
+
+## Operands and VRegs
+
+Every instruction operates on values by way of `Operand`s. An operand
+consists of the following fields:
+
+- VReg, or virtual register. *Every* operand mentions a virtual
+  register, even if it is constrained to a single physical register in
+  practice. This is because we track liveranges uniformly by vreg.
+  
+- Policy, or "constraint". Every reference to a vreg can apply some
+  constraint to the vreg at that point in the program. Valid policies are:
+  
+  - Any location;
+  - Any register of the vreg's class;
+  - Any stack slot;
+  - A particular fixed physical register; or
+  - For a def (output), a *reuse* of an input register.
+  
+- The "kind" of reference to this vreg: Def, Use, Mod. A def
+  (definition) writes to the vreg, and disregards any possible earlier
+  value. A mod (modify) reads the current value then writes a new
+  one. A use simply reads the vreg's value.
+  
+- The position: before or after the instruction.
+  - Note that to have a def (output) register available in a way that
+    does not conflict with inputs, the def should be placed at the
+    "before" position. Similarly, to have a use (input) register
+    available in a way that does not conflict with outputs, the use
+    should be placed at the "after" position.
+
+This operand-specification design allows for SSA and non-SSA code (see
+section below for details).
+
+VRegs, or virtual registers, are specified by an index and a register
+class (Float or Int). The classes are not given separately; they are
+encoded on every mention of the vreg. (In a sense, the class is an
+extra index bit, or part of the register name.) The input function
+trait does require the client to provide the exact vreg count,
+however.
+
+Implementation note: both vregs and operands are bit-packed into
+u32s. This is essential for memory-efficiency. As a result of the
+operand bit-packing in particular (including the policy constraints!),
+the allocator supports up to 2^20 (1M) vregs per function, and 2^5
+(32) physical registers per class. Later we will also see a limit of
+2^20 (1M) instructions per function. These limits are considered
```

Comment:
I wonder if the deep-vector benchmark of rustc exceeds this limit.

Reply:
That's a good question (I'll be curious for your feedback on this). That said, if there is such a large function, I'd be curious what the compile time is like, and whether it is reasonable in the first place.

I realize it's not a fully implemented answer now, but I wonder if it would make sense to have a fallback "simple greedy explicitly linear time" allocator when certain implementation limits are exceeded.

I do think it would be unfortunate to pessimize the common case to push implementation limits further to account for extreme-case benchmarks, but it's possible if we really need to. (The two places that would have to change are `Operand`, probably `u32` -> `u64`; and the sort key for half-moves, probably `u64` -> `u128`.)

Reply:
If I read it correctly the wall-time for deep-vector with optimizations enabled on perf.rust-lang.org is ~3s.

Reply:
For deep-vector the rust source is 8024 lines of which 8002 are code according to tokei. Excluding comments and empty lines the clif ir for the main function is 815846 lines after optimizations. The vcode is 543920 lines. It takes 2.6s to compile with cg_clif on my system. So while 1 million instructions may seem excessive, I do think that there is code in the wild that reaches it.

Just an idea, but may it would be possible to make the register allocator generic over the operand and sort key types? This way you can use small types when the function is small enough and (transparently) use larger types for huge functions. A fallback linear time allocator would likely also work of course.

Reply:
It takes 1.7s without optimizations and 4s with optimizations when using cg_llvm on my system by the way.

Reply:
Webkit did something similar, but the other way around: they use `u16` to index virtual registers for "small" functions with fewer than 65536 virtual registers, and otherwise fall back to an implementation that uses `u32`.

The actual implementation is generic over the integer type used: https://github.com/WebKit/WebKit/commit/28c3e585116ab

Reply:
Ah, that's neat!

We'd have to think a bit about how to handle this at the interface, given that `Operand` is exposed. Maybe we just always accept a larger (`u64`-sized) `Operand` and internally translate to a half-sized one when we copy `Operand`s to `Use` data.

I'm actually curious now what the regression will be if we just unconditionally use the larger sizes; if it's not terrible then maybe it's better to avoid the complexity. I'll experiment a bit with this.

Reply:
A quick data point: when compiling `clang.wasm`, altering `Operand` to be `u64`-sized and halfmove keys to be `u128`-sized increases compile time from 7.5s to 7.7s mean; that's about 3%. Trying just one change at a time, it seems most of the increase comes from larger `Operand`s. This benchmark is the extreme case and the number is still better than existing (BTRA) at 9.4s, but that's a nontrivial increase given that the compile-time improvements where they exist are more often 5-10%-ish. So, if we do something here, it would have to be adaptive.

Given the effort needed to do that, I think I'd prefer for the first iteration to be as-is, with the limits, and then work on this in a followup PR. The design I have in mind would be to keep `Operand` the public-facing type, enlarge it to a `u64` internally and allow for a big vreg range (32 bits?), and implement a `SmallOperand` type and an `OperandOps` trait impl'd by both internally. We always accept "big operands" but, if the vreg count is small enough, we copy to `SmallOperand`s in our internal data structures. I think the separate-impls-and-trait approach is the way to go, rather than straight parameterization, because we need custom bitpacking code for a given width.

The halfmove sort key can probably just become a `u128` to allow for large instruction ranges. The bit count on that key is `1 + 2*log2(max_blocks) + log2(max_vregs)`; i.e. it has two block-index fields and a vreg field, plus one bit. A `u128` would allow for 32-bit ranges for blocks and vregs with room to spare (and keeping these within 32 bits will allow for efficiency elsewhere).

Thoughts?